### ISSUE TITLE
CoqIDE: Unicode characters replaced with hex code on save

### ISSUE THREAD

**Issue by MSoegtropIMC** [coq/coq#11526]
When I open a .v file containing unicode characters in CoqIDE and modify + save the file, the resulting file has all unicode characters replaced with hex codes. Attached is a sample file before and after. This is very inconvenient since it produces many changes in git - I have to run sed scripts over files before commit.

I tested this on OSX, so it might be OSX specific.

[CoqIDE_Unicode_before.txt](https://github.com/coq/coq/files/4160320/CoqIDE_Unicode_before.txt)
[CoqIDE_Unicode_after.txt](https://github.com/coq/coq/files/4160321/CoqIDE_Unicode_after.txt)

**Comment by MSoegtropIMC** [comment]
Btw.: It doesn't look like coqc or CoqIDE can read the such modified files, so this is fairly serious.

**Comment by Zimmi48** [comment]
This does look OSX specific (I just tested this on my Linux machine).

**Comment by MSoegtropIMC** [comment]
Thanks for testing! I guess it since I haven't seen it on Windows either.

**Comment by llelf** [comment]
> This does look OSX specific (I just tested this on my Linux machine).

@Zimmi48 it works fine for me (8.10.1) on macOS.

My guess would be something locale-dependent (missing locales?), but after a couple of tries I have not been able to reproduce it.

**Comment by MSoegtropIMC** [comment]
I tested this on master of a few weeks back. I will make a build of latest master and test again. I think I haven't seen issues with 8.10 and 8.11, but it is not that common that files contain unicode characters.

**Comment by berpeti** [comment]
I also encountered this issue using CoqIDE (v8.13.2) in Nix (v2.3.14).

**Comment by SkySkimmer** [comment]
If you run coqide with -debug in a terminal and save the "before" file, does the debug output say something about "UTF-8 is enforced" or "Locale is (something)" or "Manual charset is (something)"? There might also be "error in transcoding".

**Comment by berpeti** [comment]
You are right, there is one such message: `[DEBUG] Locale is ANSI_X3.4-1968`.

**Comment by SkySkimmer** [comment]
So we're deliberatly converting when saving https://github.com/coq/coq/blob/aad935ddd4b574672696a71b4f002c6a2451311c/ide/coqide/ideutils.ml#L245-L247

When you open the "after" file, does the debug log contain "Converting from (something)" or "Input is UTF-8"?
If the later then I guess the problem is that we only convert on load when we get non-valid utf8 (https://github.com/coq/coq/blob/aad935ddd4b574672696a71b4f002c6a2451311c/ide/coqide/ideutils.ml#L203-L206) so if the locale encoding can pass as utf8 we get confused.

Or maybe we should just always save as utf8 regardless of locale.

**Comment by berpeti** [comment]
Yes, "Input is UTF-8" is displayed.

**Comment by SkySkimmer** [comment]
In coqide Edit -> Preferences -> Files -> File charset encoding, if you change it to UTF-8 it should prevent the bug.

**Comment by ppedrot** [comment]
So is this a real bug or just an issue with the default option?

**Comment by berpeti** [comment]
I think it is an issue with the default option.

**Comment by SkySkimmer** [comment]
I think we should change the default to UTF-8 (so it won't convert when saving) and change do_convert to always convert when the option is set to LOCALE.

**Comment by RalfJung** [comment]
This is still a problem 5 years later, corrupting people's files. Is there any plan for at least some sort of work-around to mitigate the issue?

**Comment by MSoegtropIMC** [comment]
@RalfJung : as mentioned above there is this work around:

In coqide Edit -> Preferences -> Files -> File charset encoding, if you change it to UTF-8 it should prevent the bug.

But I still think that whatever the settings are, CoqIDE shouldn't write files it neither itself nor coqc can read, whatever the settings are.

**Comment by RalfJung** [comment]
@SkySkimmer thanks for making a PR!

> as mentioned above there is this work around:

I was not sufficiently clear: by "work-around" I meant "on-by-default work-around", not "a work-around users can enable after this corrupted their file".
